HCIDOCS-188: Scaling UPI cluster with BMO doesn't explain host networ… (#77264)
openshift/openshift-docs: modules/upi-provisioning-new-hosts-in-a-upi-cluster.adoc

[NOTE]
====

> Provisioning new bare-metal hosts to the cluster using the BMO sets the `spec.externallyProvisioned` specification in the `BareMetalHost` custom resource to `false` by default.

Reviewer: I think my [previous comment](https://github.com/openshift/openshift-docs/pull/77264#discussion_r1798618089) got lost. It's critical to point out that the user **must not** set `spec.externallyProvisioned` to `true`, which is a common thing that happens.

> When provisioning bare-metal hosts to the cluster using the BMO, the `spec.externallyProvisioned` specification in the `BareMetalHost` custom resource is set to `false` by default.

Reviewer 2: ```suggestion
When provisioning new bare-metal hosts to the cluster using the BMO, the `spec.externallyProvisioned` specification in the `BareMetalHost` custom resource is set to `false` by default.
```

Author: Updated As Suggested.

> When provisioning new bare-metal hosts to the cluster using the BMO, the `spec.externallyProvisioned` specification in the `BareMetalHost` custom resource is set to `false` by default.

Reviewer 3: Perhaps "When you provision new bare-metal hosts..." 

OR

Provisioning new bare-metal hosts to the cluster using the BMO sets the `spec.externallyProvisioned` specification in the `BareMetalHost` custom resource to `false` by default.

Otherwise the 'the `spec.externallyProvisioned` specification becomes the object (i.e., doing the provisioning) and it is not really grammatically correct.

Reviewer: It's correct that the default is `false`.

It's also the case that a lot of people copy-and-paste the BMH from one of the control plane nodes (for which `externallyProvisioned` is explicitly set to `true`) and modify only the BMC address, and are then surprised when nothing happens.

That's the reason this note was here, along with the explicit callout below.

Author: Updated as suggested.

> Provisioning new bare-metal hosts to the cluster using the BMO sets the `spec.externallyProvisioned` specification in the `BareMetalHost` custom resource to `false` by default. Do not set the `spec.externallyProvisioned` specification to `true`, as this can result in unexpected behavior.

Reviewer 4: Should avoid "new" unless you really need it here :)

```suggestion
Provisioning bare-metal hosts to the cluster using the BMO sets the `spec.externallyProvisioned` specification in the `BareMetalHost` custom resource to `false` by default. Do not set the `spec.externallyProvisioned` specification to `true`, as this can result in unexpected behavior.
```

Author: Updated as suggested.

.Procedure

. Create a configuration file for the bare-metal node. Depending on whether you are using a static configuration or a DHCP server, use one of the following example `bmh.yaml` files, replacing values in the YAML to match your environment:
+
.Static configuration `bmh.yaml`
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-network-config-secret # <1>
  namespace: openshift-machine-api
type: Opaque
stringData:
 nmstate: | # <2>
  interfaces: # <3>
  - name: <nic1_name> # <4>
    type: ethernet
    state: up
    ipv4:
      address:
      - ip: <ip_address> # <5>
        prefix-length: 24
      enabled: true
  dns-resolver:
    config:
      server:
      - <dns_ip_address> # <6>
  routes:
    config:
    - destination: 0.0.0.0/0
      next-hop-address: <next_hop_ip_address> # <7>
      next-hop-interface: <next_hop_nic1_name> # <8>
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret # <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> # <9>

>   password: <base64_of_pwd> # <9>

Reviewer 4: Reuse of callouts is not supported on docs.redhat.com. I see a few in this PR - won't comment them all but please go through and make sure all are unique.

---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: <openshift-worker-<num> # <1>
  namespace: openshift-machine-api
spec:
  online: true
  bootMACAddress: <nic1_mac_address> # <10>
  bmc:
    address: <protocol>://<bmc_url> # <11>
    credentialsName: openshift-worker-<num>-bmc-secret # <1>
    disableCertificateVerification: false
  customDeploy:
    method: install_coreos
  userData:
    name: worker-user-data-managed
    namespace: openshift-machine-api
  rootDeviceHints:
    deviceName: <root_device_hint> # <12>
  preprovisioningNetworkDataName: openshift-worker-<num>-network-config-secret # <1>
----
+
--

> <1> Replace `<num>` with a unique worker number for the bare-metal nodes in the `name`, `credentialsName`, and `preprovisioningNetworkDataName` fields.

Reviewer 4: ```suggestion
<1> Replace `<num>` with a unique compute node number for the bare-metal nodes in the `name`, `credentialsName`, and `preprovisioningNetworkDataName` fields.
```

Author: Updated as suggested.

<2> Add the NMState YAML syntax to configure the host interfaces. To configure the network interface for a newly created node, specify the name of the secret that has the network configuration. Follow the `nmstate` syntax to define the network configuration for your node. See "Preparing the bare metal node" for details on configuring NMState syntax.
<3> Optional: If you have configured the network interface with `nmstate`, and you want to disable an interface, set `state: up` with the IP addresses set to `enabled: false` as shown in the following example:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-network-config-secret
  namespace: openshift-machine-api
 # ...

> interfaces:

Reviewer 4: This `interfaces` is not indented that same level as the `interfaces` stanza in the main YAML block. Is this showing an alternate configuration of the whole first section? 

* **IF** the `# ...` on line 102 is hiding structure, it needs to be fully written out to get to `interfaces`
* **IF** the indent is right and line 102 is _not_ hiding structure, consider making this a separate config file example. Putting this much in a callout is hard to follow.
* in any case, I would suggest making these different config examples substeps of step one, i.e.

1. Create a configuration file for the bare-metal node.
  * If you are using a static configuration, use the following example `bmh.yaml` file, replacing values in the YAML to match your environment:
  [YAML block for static]
  * (Possible new bullet for the `state: up` / `enabled: false` flavor?)
  [YAML block for it]
  * If you are using a DHCP server, use the following example `bmh.yaml` file, replacing values in the YAML to match your environment:
  [YAML block for DHCP]

I am not sure how well I have explained what I am trying to say via GH comment, I would be happy to chat with you realtime if that would be more helpful :sweat_smile:

<11> Replace `<protocol>` with the BMC protocol, such as IPMI, Redfish, or others. Replace `<bmc_url>` with the URL of the bare-metal node's baseboard management controller.
<12> Optional: Replace `<root_device_hint>` with a device path when specifying a root device hint. See "Root device hints" for additional details.
--
+
.DHCP configuration `bmh.yaml`:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret # <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> # <2>
  password: <base64_of_pwd> # <2>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> # <1>
  namespace: openshift-machine-api
spec:
  online: true
  bootMACAddress: <nic1_mac_address> # <3>
  bmc:
    address: <protocol>://<bmc_url> # <4>
    credentialsName: openshift-worker-<num>-bmc
    disableCertificateVerification: false
  customDeploy:
    method: install_coreos
  userData:
    name: worker-user-data-managed
    namespace: openshift-machine-api
  rootDeviceHints:
    deviceName: <root_device_hint> # <5>
----
+

> <1> Replace `<num>` with a unique worker number for the bare-metal nodes in the `name` and `credentialsName` fields.

Reviewer 4: ```suggestion
<1> Replace `<num>` with a unique compute node number for the bare-metal nodes in the `name` and `credentialsName` fields.
```

+
<2> Replace `<base64_of_uid>` and `<base64_of_pwd>` with the base64 string of the user name and password.
+
<3> Replace `<nic1_mac_address>` with the MAC address of the bare metal node's first NIC. See the "BMC addressing" section for additional BMC configuration options.
+
<4> Replace `<protocol>` with the BMC protocol, such as IPMI, RedFish, or others. Replace `<bmc_url>` with the URL of the bare metal node's baseboard management controller.
+
<5> To skip certificate validation, set `disableCertificateVerification` to true.
+
<6> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint.
+
[NOTE]
====
If the MAC address of an existing bare metal node matches the MAC address of a bare metal host that you are attempting to provision, then the installation will fail. If the host enrollment, inspection, cleaning, or other steps fail, the Bare Metal Operator retries the installation continuously. See "Diagnosing a host duplicate MAC address" for more information.
====

. Create the bare metal node:
+
[source,terminal]
----

> $ oc -n openshift-machine-api create -f bmh.yaml

Reviewer 5: nit: I don't think `oc create` requires a namespace

Author: What would be the correct command? `$ oc create -f bmh.yaml`?

Reviewer 5: Correct

Author: Updated As Suggested.

+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
 name: openshift-worker-<num>-network-config-secret <1>
 namespace: openshift-machine-api
type: Opaque
stringData:
 nmstate: | <2>
  interfaces: <3>
  - name: <nic1_name> <4>
    type: ethernet
    state: up
    ipv4:
      address:
      - ip: <ip_address> <4>
        prefix-length: 24
      enabled: true
  dns-resolver:
    config:
      server:
      - <dns_ip_address> <4>
  routes:
    config:
    - destination: 0.0.0.0/0
      next-hop-address: <next_hop_ip_address> <4>
      next-hop-interface: <next_hop_nic1_name> <4>
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <5>
  password: <base64_of_pwd> <5>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <6>
  bmc:
    address: <protocol>://<bmc_url> <7>
    credentialsName: openshift-worker-<num>-bmc-secret <1>
    disableCertificateVerification: True <8>
    username: <bmc_username> <9>

>     password: <bmc_password> <9>

Reviewer 5: username and password are not valid here

Author: Updated As Suggested.

<3> Optional: If you have configured the network interface with `nmstate`, and you want to disable an interface, set `state: up` with the IP addresses set to `enabled: false` as shown:
+
[source,yaml]
----
---
   interfaces:
   - name: <nic_name>
     type: ethernet
     state: up
     ipv4:
       enabled: false
     ipv6:
       enabled: false
----
<4> Replace `<nic1_name>`, `<ip_address>`, `<dns_ip_address>`, `<next_hop_ip_address>`, and `<next_hop_nic1_name>` with appropriate values.
<5> Replace `<base64_of_uid>` and  `<base64_of_pwd>` with the base64 string of the user name and password.
<6> Replace `<nic1_mac_address>` with the MAC address of the bare metal node's first NIC. See the "BMC addressing" section for additional BMC configuration options.
<7> Replace `<protocol>` with the BMC protocol, such as IPMI, RedFish, or others. Replace `<bmc_url>` with the URL of the bare metal node's baseboard management controller.

> <8> To skip certificate validation, set `disableCertificateVerification` to true.

Author: Suggested rewrite:
> <8> To skip TLS certificate validation, set `disableCertificateVerification` to true. Disabling TLS certificate validation is not recommended.

> <8> To skip certificate validation, set `disableCertificateVerification` to true.

Reviewer 5: LGTM

> <8> To skip certificate validation, set `disableCertificateVerification` to true.

Author: Updated As Suggested.

<9> Replace `<bmc_username>` and `<bmc_password>` with the string of the BMC user name and password.

> <10> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint.

Reviewer 5: This field has many variants, the device path is the least recommended one. Do we have any explicit root device hints documentation anywhere to link to?

Author: @dtantsur, is this the [documentation](https://docs.redhat.com/en/documentation/openshift_container_platform/4.16/html/installing/deploying-installer-provisioned-clusters-on-bare-metal#root-device-hints_ipi-install-installation-workflow) you're referring to?

Reviewer 5: Yes, exactly. Maybe link to it?

Author: Updated As Suggested.

Reviewer 2: I cant find the link to the root device hint documentation

--

* *DHCP configuration* `bmh.yaml`:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <2>
  password: <base64_of_pwd> <2>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <3>
  bmc:
    address: <protocol>://<bmc_url> <4>
    credentialsName: openshift-worker-<num>-bmc-secret <1>
    disableCertificateVerification: True <5>
    username: <bmc_username> <6>

>     password: <bmc_password> <6>

Reviewer 5: Same re username/password

Author: Updated As Suggested.

  rootDeviceHints:
    deviceName: <root_device_hint> <7>

>   preprovisioningNetworkDataName: openshift-worker-<num>-network-config-secret

Reviewer 5: This secret is not present in the DHCP example.

Author: Updated As Suggested.

. Create the `Secret` CR and the `BareMetalHost` CR.

> . Create a configuration file for the bare metal node. Depending on whether you are using a static configuration or a DHCP server, use one of the following example `bmh.yaml` files, replacing values in the YAML to match your environment:

Author: Updated As Suggested.

 nmstate: | <3>
  interfaces: <4>
  - name: <nic1_name> <5>
    type: ethernet
    state: up
    ipv4:
      address:
      - ip: <ip_address> <5>
        prefix-length: 24
      enabled: true
  dns-resolver:
    config:
      server:
      - <dns_ip_address> <5>
  routes:
    config:
    - destination: 0.0.0.0/0
      next-hop-address: <next_hop_ip_address> <5>
      next-hop-interface: <next_hop_nic1_name> <5>
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <2>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <6>
  password: <base64_of_pwd> <6>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> <2>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <7>
  bmc:
    address: <protocol>://<bmc_url> <8>
    credentialsName: openshift-worker-<num>-bmc-secret <2>
    disableCertificateVerification: True <9>
    username: <bmc_username> <10>
    password: <bmc_password> <10>
  rootDeviceHints:
    deviceName: <root_device_hint> <11>
  preprovisioningNetworkDataName: openshift-worker-<num>-network-config-secret <12>
----
+
--

> <1> To configure the network interface for a newly created node, specify the name of the secret that contains the network configuration. Follow the `nmstate` syntax to define the network configuration for your node. See "Optional: Configuring host network interfaces in the install-config.yaml file" for details on configuring NMState syntax.

Author: [Configuring host network interfaces](https://docs.redhat.com/en/documentation/openshift_container_platform/4.15/html/installing/deploying-installer-provisioned-clusters-on-bare-metal#configuring-host-network-interfaces-in-the-install-config-yaml-file_ipi-install-installation-workflow) is in the` installing_bare_metal_ipi/ipi-install-installation-workflow.adoc` assembly.

> <1> To configure the network interface for a newly created node, specify the name of the secret that contains the network configuration. Follow the `nmstate` syntax to define the network configuration for your node. See "Optional: Configuring host network interfaces in the install-config.yaml file" for details on configuring NMState syntax.

Author: Updated As Suggested.

> <2> Replace `<num>` for the worker number of the bare metal node in the `name` fields, the `credentialsName` field, and the `preprovisioningNetworkDataName` field.

Author: Updated `preprovisioningNetworkDataName` call out. I'll verify the numbers associated with the worker nodes.

> <2> Replace `<num>` for the worker number of the bare metal node in the `name` fields, the `credentialsName` field, and the `preprovisioningNetworkDataName` field.

Author: @dtantsur, could you please confirm if there is any specific number assigned to the worker node or if it's actually the complete count/total number of worker nodes?

> <2> Replace `<num>` for the worker number of the bare metal node in the `name` fields, the `credentialsName` field, and the `preprovisioningNetworkDataName` field.

Author: [Per](https://github.com/openshift/openshift-docs/pull/77264/files#r1665386362) Dmitry:

> These examples assume a specific naming schema where each worker has a sequential number. In fact, the BMH names can be completely arbitrary.

> <2> Replace `<num>` for the worker number of the bare metal node in the `name` fields, the `credentialsName` field, and the `preprovisioningNetworkDataName` field.

Author: [Per](https://github.com/openshift/openshift-docs/pull/77264/files#r1665386362) Dmitry:

> These examples assume a specific naming schema where each worker has a sequential number. In fact, the BMH names can be completely arbitrary.

> <3> Add the NMState YAML syntax to configure the host interfaces.

Author: Updated As Suggested.

> <4> Optional: If you have configured the network interface with `nmstate`, and you want to disable an interface, set `state: up` with the IP addresses set to `enabled: false` as shown:

Author: Updated As Suggested.

+
[source,yaml]
----
---
   interfaces:
   - name: <nic_name>
     type: ethernet
     state: up
     ipv4:
       enabled: false
     ipv6:
       enabled: false
----
<5> Replace `<nic1_name>`, `<ip_address>`, `<dns_ip_address>`, `<next_hop_ip_address>` and `<next_hop_nic1_name>` with appropriate values.
<6> Replace `<base64_of_uid>` and  `<base64_of_pwd>` with the base64 string of the user name and password.

> <7> Replace `<nic1_mac_address>` with the MAC address of the bare metal node's first NIC. See the "BMC addressing" section for additional BMC configuration options.

Author: BMC Addressing (/installing/installing_bare_metal_ipi/ipi-install-installation-workflow.adoc#bmc-addressing_ipi-install-installation-workflow) looks like an IPI step, but this procedure is for UPI. I'll verify if the IPI content can be reused for this UPI content.

> <7> Replace `<nic1_mac_address>` with the MAC address of the bare metal node's first NIC. See the "BMC addressing" section for additional BMC configuration options.

Author: @dtantsur, could you please confirm if this specific IPI content can be reused with UPI content?

<8> Replace `<protocol>` with the BMC protocol, such as IPMI, RedFish, or others. Replace `<bmc_url>` with the URL of the bare metal node's baseboard management controller.

> <9> To skip certificate validation, set `disableCertificateVerification` to true.

Author: Updated As Suggested. I would have thought it was `true`, but the docs I reviewed list `True` in the examples and callouts.

<11> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint.
<12> Optional: If you have configured the network interface for the newly created node, provide the network configuration secret name in the `preprovisioningNetworkDataName` of the BareMetalHost CR.
--

* *DHCP configuration* `bmh.yaml`:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <2>
  password: <base64_of_pwd> <2>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <3>
  bmc:
    address: <protocol>://<bmc_url> <4>
    credentialsName: openshift-worker-<num>-bmc-secret <1>
    disableCertificateVerification: True <5>
    username: <bmc_username> <6>
    password: <bmc_password> <6>
  rootDeviceHints:
    deviceName: <root_device_hint> <7>
  preprovisioningNetworkDataName: openshift-worker-<num>-network-config-secret <8>
----
+

> <1> Replace `<num>` for the worker number of the bare metal node in the `name` fields, the `credentialsName` field, and the `preprovisioningNetworkDataName` field.

Author: I'll verify the numbers associated with the worker nodes.

> <1> Replace `<num>` for the worker number of the bare metal node in the `name` fields, the `credentialsName` field, and the `preprovisioningNetworkDataName` field.

Author: @dtantsur, could you please confirm if there is any specific number assigned to the worker node or if it's actually the complete count/total number of worker nodes?

> <1> Replace `<num>` for the worker number of the bare metal node in the `name` fields, the `credentialsName` field, and the `preprovisioningNetworkDataName` field.

Author: [Per](https://github.com/openshift/openshift-docs/pull/77264/files#r1665386362) Dmitry:

> These examples assume a specific naming schema where each worker has a sequential number. In fact, the BMH names can be completely arbitrary.

+
<2> Replace `<base64_of_uid>` and `<base64_of_pwd>` with the base64 string of the user name and password.
+
<3> Replace `<nic1_mac_address>` with the MAC address of the bare metal node's first NIC. See the "BMC addressing" section for additional BMC configuration options.
+
<4> Replace `<protocol>` with the BMC protocol, such as IPMI, RedFish, or others. Replace `<bmc_url>` with the URL of the bare metal node's baseboard management controller.
+

> <5> To skip certificate validation, set `disableCertificateVerification` to true.

Author: Updated As Suggested. I would have thought it was `true`, but the docs I reviewed list `True` in the examples and callouts.

> <5> To skip certificate validation, set `disableCertificateVerification` to true.

Reviewer 6: Same as https://github.com/openshift/openshift-docs/pull/77264/files#diff-e880e9c1bb4d1f0174b865aa603e0f69e243d82008ac0f4b1f7c18e127996d4aR111. Do we want the _Disabling not recommended_ wording here also?

+
<6> Replace `<bmc_username>` and `<bmc_password>` with the string of the BMC user name and password.
+
<7> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint.
+

> <8> Optional: If you have configured the network interface for the newly created node, provide the network configuration secret name in the `preprovisioningNetworkDataName` of the BareMetalHost CR.

Author: Updated As Suggested.

+
[NOTE]
====

> If the MAC address of an existing bare metal node matches the MAC address of a bare metal host that you are attempting to provision, then the Ironic installation will fail. If the host enrollment, inspection, cleaning, or other Ironic steps fail, the Bare Metal Operator retries the installation continuously. See "Diagnosing a host duplicate MAC address" for more information.

Author: I'll verify the Ironic install details.

> If the MAC address of an existing bare metal node matches the MAC address of a bare metal host that you are attempting to provision, then the Ironic installation will fail. If the host enrollment, inspection, cleaning, or other Ironic steps fail, the Bare Metal Operator retries the installation continuously. See "Diagnosing a host duplicate MAC address" for more information.

Author: [Diagnosing a host duplicate MAC address](https://docs.redhat.com/en/documentation/openshift_container_platform/4.15/html/installing/deploying-installer-provisioned-clusters-on-bare-metal#ipi-install-diagnosing-duplicate-mac-address_ipi-install-expanding)

> If the MAC address of an existing bare metal node matches the MAC address of a bare metal host that you are attempting to provision, then the Ironic installation will fail. If the host enrollment, inspection, cleaning, or other Ironic steps fail, the Bare Metal Operator retries the installation continuously. See "Diagnosing a host duplicate MAC address" for more information.

Reviewer 5: Do we have enough context on "Ironic installation" and "Ironic steps"? I'm afraid we don't explain what Ironic is in this document, I'd avoid referring to it. It seems like you can safely remove "Ironic" without losing anything from the text.

> If the MAC address of an existing bare metal node matches the MAC address of a bare metal host that you are attempting to provision, then the Ironic installation will fail. If the host enrollment, inspection, cleaning, or other Ironic steps fail, the Bare Metal Operator retries the installation continuously. See "Diagnosing a host duplicate MAC address" for more information.

Author: Updated As Suggested.

====

> . Create the bare metal node:

Author: @dtantsur could you please confirm if the above steps regarding bare metal host and bare metal node are correct? Is the `oc apply` command required in order to apply configuration changes?

> . Create the bare metal node:

Reviewer 5: I'm confused. This document has two different examples of `bmh.yaml`: one with explicit configuration and one using DHCP. They are independent, right? You probably need to clarify that and add an `oc create` command for each.

> . Create the bare metal node:

Author: Updated As Suggested.

+
[source,terminal]
----
$ oc -n openshift-machine-api create -f bmh.yaml
----
+
.Example output
[source,terminal]
----
secret/openshift-worker-<num>-network-config-secret created
secret/openshift-worker-<num>-bmc-secret created
baremetalhost.metal3.io/openshift-worker-<num> created
----
+

> Where `<num>` will be the worker number.

Reviewer 5: These examples assume a specific naming schema where each worker has a sequential number. In fact, the BMH names can be completely arbitrary.

> . Power up and inspect the bare metal node:

Reviewer 7: IBM style guide suggests to not use `power up` `https://www.ibm.com/docs/en/ibm-style?topic=word-usage` 
Do not use. Use power on or turn on.

What do we inspect in the bare metal node? Any specific details??

Reviewer 5: Inspect = boot the node and fetch hardware details, essentially.

Author: Updated As Suggested.

+
[source,terminal]
----
$ oc -n openshift-machine-api get bmh openshift-worker-<num>
----
+

> Where `<num>` is the worker node number.

Author: @dtantsur, could you please confirm if there is any specific number assigned to the worker node or if it's actually the complete count/total number of worker nodes?

> Where `<num>` is the worker node number.

Author: [Per](https://github.com/openshift/openshift-docs/pull/77264/files#r1665386362) Dmitry:

> These examples assume a specific naming schema where each worker has a sequential number. In fact, the BMH names can be completely arbitrary.

> Where `<num>` is the worker node number.

Author: [Per](https://github.com/openshift/openshift-docs/pull/77264/files#r1665386362) Dmitry:

> These examples assume a specific naming schema where each worker has a sequential number. In fact, the BMH names can be completely arbitrary.

> Where `<num>` is the worker node number.

Author: [Per](https://github.com/openshift/openshift-docs/pull/77264/files#r1665386362) Dmitry:

> These examples assume a specific naming schema where each worker has a sequential number. In fact, the BMH names can be completely arbitrary.

+
.Example output
[source,terminal]
----
NAME                    STATE       CONSUMER   ONLINE   ERROR
openshift-worker-<num>  available              true
----
+
[NOTE]

> ====
> To allow the worker node to join the cluster, scale the `machineset` object to the number of the `BareMetalHost` objects. You can scale nodes either manually or automatically. To scale nodes automatically, use the `metal3.io/autoscale-to-hosts` annotation for `machineset`.
> ====

Reviewer 7: ```suggestion
====
To add the worker node to the cluster, scale the `machineset` object to the number of the `BareMetalHost` objects. You can scale nodes either manually or automatically. To scale nodes automatically, use the `metal3.io/autoscale-to-hosts` annotation for the `machineset` object.
====
```
Avoid instances of anthropomorphism.

Author: Updated As Suggested.

+

> . Create the bare-metal host object by running the following command:

Reviewer 7: Here, we are again creating a bare metal host. Though we have already provided the configuration details in existing step 2. the procedure looks confusing.

Author: @dtantsur could you please confirm if the above steps regarding bare metal host and bare metal node are correct?

Reviewer 5: There is no need to run `oc create` twice.

Author: Updated As Suggested.

 namespace: openshift-machine-api
type: Opaque
stringData:
 nmstate: | <2>
  interfaces: <3>
  - name: <nic1_name> <4>
    type: ethernet
    state: up
    ipv4:
      address:
      - ip: <ip_address> <4>
        prefix-length: 24
      enabled: true
  dns-resolver:
    config:
      server:
      - <dns_ip_address> <4>
  routes:
    config:
    - destination: 0.0.0.0/0
      next-hop-address: <next_hop_ip_address> <4>
      next-hop-interface: <next_hop_nic1_name> <4>
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <5>
  password: <base64_of_pwd> <5>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <6>
  bmc:
    address: <protocol>://<bmc_url> <7>
    credentialsName: openshift-worker-<num>-bmc-secret <1>
    disableCertificateVerification: True <8>
  rootDeviceHints:
    deviceName: <root_device_hint> <9>
  preprovisioningNetworkDataName: openshift-worker-<num>-network-config-secret <1>
----
+
--

> <1> Replace `<num>` for the machine and worker numbers of the bare-metal nodes in the `name`, `credentialsName`, and the `preprovisioningNetworkDataName` fields.

Reviewer 6: > for the machine and worker numbers of the bare-metal nodes 

I don't know what this means. (Does each bare metal node have an identifying number, or does this refer to quantity??)  Will the typical user know?

Author: The `<num>` value is unique to the cluster and defined by the user.

> <2> Add the NMState YAML syntax to configure the host interfaces. To configure the network interface for a newly created node, specify the name of the secret that has the network configuration. Follow the `nmstate` syntax to define the network configuration for your node. See link:https://docs.redhat.com/en/documentation/openshift_container_platform/4.15/html/installing/deploying-installer-provisioned-clusters-on-bare-metal#configuring-host-network-interfaces-in-the-install-config-yaml-file_ipi-install-installation-workflow[Optional: Configuring host network interfaces] for details on configuring NMState syntax.

Reviewer 6: No links in modules (unless your team has better rules!)
```suggestion
<2> Add the NMState YAML syntax to configure the host interfaces. To configure the network interface for a newly created node, specify the name of the secret that has the network configuration. Follow the `nmstate` syntax to define the network configuration for your node. See _Optional: Configuring host network interfaces_ for details on configuring NMState syntax.
```

Author: Updated As Suggested.

<3> Optional: If you have configured the network interface with `nmstate`, and you want to disable an interface, set `state: up` with the IP addresses set to `enabled: false` as shown:
+
[source,yaml]
----
---

>    interfaces:
>    - name: <nic_name>
>      type: ethernet
>      state: up
>      ipv4:
>        enabled: false
>      ipv6:
>        enabled: false

Reviewer 6: Code [blocks take metadata](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#required-fields) and [`# ...` when snipping the code](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#signs-and-symbols-in-code-blocks).

```suggestion
apiVersion: v1
kind: Secret
metadata:
 name: openshift-worker-<num>-network-config-secret 
 namespace: openshift-machine-api
 # ...
   interfaces:
   - name: <nic_name>
     type: ethernet
     state: up
     ipv4:
       enabled: false
     ipv6:
       enabled: false
# ...
```

Author: Updated As Suggested.

----
<4> Replace `<nic1_name>`, `<ip_address>`, `<dns_ip_address>`, `<next_hop_ip_address>`, and `<next_hop_nic1_name>` with appropriate values.
<5> Replace `<base64_of_uid>` and  `<base64_of_pwd>` with the base64 string of the user name and password.
<6> Replace `<nic1_mac_address>` with the MAC address of the bare-metal node's first NIC. See the "BMC addressing" section for additional BMC configuration options.

> <7> Replace `<protocol>` with the BMC protocol, such as IPMI, RedFish, or others. Replace `<bmc_url>` with the URL of the bare-metal node's baseboard management controller.

Reviewer 8: Typo: Change RedFish to Redfish.

Author: Updated As Suggested.

> <8> Optional: To skip TLS certificate validation, set `disableCertificateVerification` to true. Disabling TLS certificate validation is not recommended.

Reviewer 6: ```suggestion
<8> Optional: To skip TLS certificate validation, set `disableCertificateVerification` to `true`. Disabling TLS certificate validation is not recommended.
```
or is it
```suggestion
<8> Optional: To skip TLS certificate validation, set `disableCertificateVerification` to `True`. Disabling TLS certificate validation is not recommended.
```

Author: Updated As Suggested with `true`.

<9> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint.
--

* *DHCP configuration* `bmh.yaml`:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <2>
  password: <base64_of_pwd> <2>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <3>
  bmc:
    address: <protocol>://<bmc_url> <4>
    credentialsName: openshift-worker-<num>-bmc-secret <1>
    disableCertificateVerification: True <5>
  rootDeviceHints:
    deviceName: <root_device_hint> <6>
----

> +

Reviewer 6: Do we need the `+` characters in this list?

Author: I don't believe so. Removed `+` and confirmed proper formatting via preview.

> <1> Replace `<num>` for the machine and worker numbers of the bare-metal nodes in the `name`, `credentialsName`, and the `preprovisioningNetworkDataName` fields.

Reviewer 6: If you made changes to the [same sentence above](https://github.com/openshift/openshift-docs/pull/77264/files#diff-e880e9c1bb4d1f0174b865aa603e0f69e243d82008ac0f4b1f7c18e127996d4aR91), also here.

Author: Updated As Suggested.

+
<2> Replace `<base64_of_uid>` and `<base64_of_pwd>` with the base64 string of the user name and password.
+
<3> Replace `<nic1_mac_address>` with the MAC address of the bare-metal node's first NIC. See the "BMC addressing" section for additional BMC configuration options.
+

> <4> Replace `<protocol>` with the BMC protocol, such as IPMI, RedFish, or others. Replace `<bmc_url>` with the URL of the bare-metal node's baseboard management controller.

Reviewer 8: Fix RedFish typo

Author: Updated As Suggested.

+
<5> To skip certificate validation, set `disableCertificateVerification` to true.
+

> <6> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint. See link:https://docs.redhat.com/en/documentation/openshift_container_platform/<latest>/html/installing/deploying-installer-provisioned-clusters-on-bare-metal#root-device-hints_ipi-install-installation-workflow[Root device hints] for additional details.

Reviewer 6: No internal links in modules

Author: Updated As Suggested.

<8> Optional: To skip TLS certificate validation, set `disableCertificateVerification` to `true`. Disabling TLS certificate validation is not recommended.
<9> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint.
--

* *DHCP configuration* `bmh.yaml`:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: worker1-bmc
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid>
  password: <base64_of_pwd>
  username: <base64_of_uid> <2>
  password: <base64_of_pwd> <2>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: worker1
  name: openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <3>
  bmc:
    address: <protocol>://<bmc_url> <1>
    credentialsName: "worker1-bmc"
  bootMACAddress: <nic1_mac_address>
  externallyProvisioned: false <2>
  customDeploy:
    method: install_coreos
  online: true
  userData:
    name: worker-user-data-managed

>     namespace: openshift-machine-api

Reviewer 2: @dtantsur is userData section not needed?

Reviewer 5: Yeah, this change is a bit problematic. We're replacing an complete a example of a worker that is ready to join the cluster with an example of a host that will become "available" and wait for further instructions. I don't think that's what users want. I think the difference between the two examples should only be in the preprovisioningNetworkDataName field.

Author: @dtantsur, thanks so much for the clarification. The `userData` section (in red) is actually already deleted from the content. 

If I'm understanding you correctly, the `preprovisioningNetworkDataName` field has a different definition or value for the DHCP section? Could you please confirm if this is the case and what that value should be? Thanks!

Author: cc: @sgoveas

Reviewer 5: preprovisioningNetworkDataName is not required in the  DHCP case because DHCP is the default behavior.

Reviewer 5: To clarify my early comment: userData most likely should not be deleted. If you remove customDeploy and userData, you will not end up with an OpenShift node, just with an available host that will be waiting for further instructions.

Author: Updated As Suggested. @dtantsur can you please confirm if this update is correct? Thank you!

Reviewer 2: There is an issue with the alignment and ordering and I think customDeploy and userData section are needed in both static and dhcp configs
```
spec:
  online: True
  bootMACAddress: <nic1_mac_address> 
  bmc:
    address: <protocol>://<bmc_url> 
    credentialsName: openshift-worker-<num>-bmc-secret
    disableCertificateVerification: True
  customDeploy:
    method: install_coreos
  online: true
  userData:
    name: worker-user-data-managed 
    namespace: openshift-machine-api
  rootDeviceHints:
    deviceName: <root_device_hint>
```

Author: Updated As Suggested.

----
---
apiVersion: v1
kind: Secret
metadata:
 name: openshift-worker-<num>-network-config-secret <1>
 namespace: openshift-machine-api
type: Opaque
stringData:
 nmstate: | <2>
  interfaces: <3>
  - name: <nic1_name> <4>
    type: ethernet
    state: up
    ipv4:
      address:
      - ip: <ip_address> <4>
        prefix-length: 24
      enabled: true
  dns-resolver:
    config:
      server:
      - <dns_ip_address> <4>
  routes:
    config:
    - destination: 0.0.0.0/0
      next-hop-address: <next_hop_ip_address> <4>
      next-hop-interface: <next_hop_nic1_name> <4>
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <5>
  password: <base64_of_pwd> <5>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <6>
  bmc:
    address: <protocol>://<bmc_url> <7>
    credentialsName: openshift-worker-<num>-bmc-secret <1>
    disableCertificateVerification: True <8>
  rootDeviceHints:
    deviceName: <root_device_hint> <9>

>   preprovisioningNetworkDataName: openshift-worker-<num>-network-config-secret <1>

Reviewer 2: Please add `userData` and `customDeploy` sections to this yaml file

Author: Updated As Suggested.

<5> Replace `<base64_of_uid>` and  `<base64_of_pwd>` with the base64 string of the user name and password.
<6> Replace `<nic1_mac_address>` with the MAC address of the bare-metal node's first NIC. See the "BMC addressing" section for additional BMC configuration options.
<7> Replace `<protocol>` with the BMC protocol, such as IPMI, Redfish, or others. Replace `<bmc_url>` with the URL of the bare-metal node's baseboard management controller.
<8> Optional: To skip TLS certificate validation, set `disableCertificateVerification` to `true`. Disabling TLS certificate validation is not recommended.
<9> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint.
--

* *DHCP configuration* `bmh.yaml`:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <2>
  password: <base64_of_pwd> <2>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <3>
  bmc:

>       customDeploy:

Reviewer 2: @kdrake20 the indentation and order needs fixing

Author: Updated As Suggested.

> . Create the `Secret` CR and the `BareMetalHost` CR.

Reviewer 5: I think this item should be removed, it's confusing without further context.

Author: Updated as suggested.

. Create a configuration file for the bare-metal node. Depending on whether you are using a static configuration or a DHCP server, use one of the following example `bmh.yaml` files, replacing values in the YAML to match your environment:
+
[source,terminal]
----
$ vim bmh.yaml
----
* *Static configuration* `bmh.yaml`:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
name: openshift-worker-<num>-network-config-secret <1>
namespace: openshift-machine-api
type: Opaque
stringData:
 nmstate: | <2>
  interfaces: <3>
  - name: <nic1_name> <4>
    type: ethernet
    state: up
    ipv4:
      address:
      - ip: <ip_address> <4>
        prefix-length: 24
      enabled: true
  dns-resolver:
    config:
      server:
      - <dns_ip_address> <4>
  routes:
    config:
    - destination: 0.0.0.0/0
      next-hop-address: <next_hop_ip_address> <4>
      next-hop-interface: <next_hop_nic1_name> <4>
---
apiVersion: v1
kind: Secret
metadata:
name: openshift-worker-<num>-bmc-secret <1>
namespace: openshift-machine-api
type: Opaque
data:
 username: <base64_of_uid> <5>
 password: <base64_of_pwd> <5>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost

> metadata:
> customDeploy:
> method: install_coreos
> online: true
> userData:
> name: <openshift-worker-<num> <1>
> namespace: openshift-machine-api
> spec:
> bootMACAddress: <nic1_mac_address> <6>
> bmc:
> address: <protocol>://<bmc_url> <7>
> credentialsName: openshift-worker-<num>-bmc-secret <1>
> disableCertificateVerification: true <8>
> rootDeviceHints:
> deviceName: <root_device_hint> <9>
> preprovisioningNetworkDataName: openshift-worker-<num>-network-config-secret

Reviewer 2: ```suggestion
metadata:
  name: <openshift-worker-<num>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> 
  bmc:
    address: <protocol>://<bmc_url> 
    credentialsName: openshift-worker-<num>-bmc-secret
    disableCertificateVerification: True
  customDeploy:
    method: install_coreos
  online: true
  userData:
    name: worker-user-data-managed 
    namespace: openshift-machine-api
  rootDeviceHints:
    deviceName: <root_device_hint>
```

Reviewer 2: Note on line 12
>To provision bare-metal hosts to the cluster by using the BMO, you must set the `spec.externallyProvisioned` specification in the `BareMetalHost` custom resource to `false`.

`must set` is confusing as we have not set it in the above bmh.yaml. `spec.externallyProvisioned`  is `false` by default if not set

Author: Updated As Suggested.

<6> Replace `<nic1_mac_address>` with the MAC address of the bare-metal node's first NIC. See the "BMC addressing" section for additional BMC configuration options.
<7> Replace `<protocol>` with the BMC protocol, such as IPMI, Redfish, or others. Replace `<bmc_url>` with the URL of the bare-metal node's baseboard management controller.
<8> Optional: To skip TLS certificate validation, set `disableCertificateVerification` to `true`. Disabling TLS certificate validation is not recommended.
<9> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint.
--

* *DHCP configuration* `bmh.yaml`:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <2>
  password: <base64_of_pwd> <2>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <3>
  bmc:
    customDeploy:
    method: install_coreos
online: true
userData:

>     name: worker-user-data-managed <1>

Reviewer 2: no `<num>` here

Author: Updated As Suggested.

    namespace: openshift-machine-api
    address: <protocol>://<bmc_url> <4>
    credentialsName: openshift-worker-<num>-bmc-secret <1>
    disableCertificateVerification: True <5>
rootDeviceHints:

>     deviceName: <root_device_hint> <6>

Reviewer 2: ```suggestion
metadata:
  name: <openshift-worker-<num> 
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> 
  bmc:
    address: <protocol>://<bmc_url> 
    credentialsName: openshift-worker-<num>-bmc-secret
    disableCertificateVerification: True
  customDeploy:
    method: install_coreos
  online: true
  userData:
    name: worker-user-data-managed 
    namespace: openshift-machine-api
  rootDeviceHints:
    deviceName: <root_device_hint>

Author: Updated As Suggested.

    method: install_coreos
  online: true
  userData:
    name: worker-user-data-managed
online: true
userData:
    name: worker-user-data-managed <1>
    namespace: openshift-machine-api
    address: <protocol>://<bmc_url> <4>
    credentialsName: openshift-worker-<num>-bmc-secret <1>
    disableCertificateVerification: True <5>
rootDeviceHints:
    deviceName: <root_device_hint> <6>
----
<1> You can only use bare-metal host drivers that support virtual media networking booting, for example `redfish-virtualmedia` and `idrac-virtualmedia`.

> <2> You must set the `spec.externallyProvisioned` specification in the `BareMetalHost` custom resource to `false`. The default value is `false`.

Reviewer 2: Note on line 12
>To provision bare-metal hosts to the cluster by using the BMO, you must set the `spec.externallyProvisioned` specification in the `BareMetalHost` custom resource to `false`.

`must set` is confusing as we have not set it in the above bmh.yaml. `spec.externallyProvisioned`  is `false` by default if not set

Author: Updated As Suggested.

Reviewer: Same as https://github.com/openshift/openshift-docs/pull/77264/files#r1800329812
This note was here to protect users against copying-and-pasting from a control-plane BMH. It doesn't matter how it is phrased:

- must be set to `false`. the default is `false`
- must be set to the default, which is `false`
- must not be set to `true`

but removing it altogether will cause an increase in support tickets.

. Create the `Secret` CR and the `BareMetalHost` CR.

. Create a configuration file for the bare-metal node. Depending on whether you are using a static configuration or a DHCP server, use one of the following example `bmh.yaml` files, replacing values in the YAML to match your environment:
+
[source,terminal]
----
$ vim bmh.yaml

> ----

Author: RE: `oc apply`, `oc create` is actually correct here, though both work.` oc create` is to create a new resource that we know does not exist. If the resource was already created, running `oc create` will throw an error. `oc apply` can create or update the resource (if it already exists). For this content, we don't want the user to update an already existing resource, hence `oc create`.

* *Static configuration* `bmh.yaml`:

> +

Author: Updated As Suggested.

    config:
      server:
      - <dns_ip_address> <4>
  routes:
    config:
    - destination: 0.0.0.0/0
      next-hop-address: <next_hop_ip_address> <4>
      next-hop-interface: <next_hop_nic1_name> <4>
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <5>
  password: <base64_of_pwd> <5>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: <openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: true
  bootMACAddress: <nic1_mac_address> <6>
  bmc:
    address: <protocol>://<bmc_url> <7>
    credentialsName: openshift-worker-<num>-bmc-secret <1>
    disableCertificateVerification: True <8>
  customDeploy:
    method: install_coreos
  userData:
    name: worker-user-data-managed
    namespace: openshift-machine-api
  rootDeviceHints:
    deviceName: <root_device_hint> <9>
  preprovisioningNetworkDataName: openshift-worker-<num>-network-config-secret <1>
----
+
--
<1> Replace `<num>` with a unique worker number for the bare-metal nodes in the `name`, `credentialsName`, and `preprovisioningNetworkDataName` fields.

> <2> Add the NMState YAML syntax to configure the host interfaces. To configure the network interface for a newly created node, specify the name of the secret that has the network configuration. Follow the `nmstate` syntax to define the network configuration for your node. See "Optional: Configuring host network interfaces" for details on configuring NMState syntax.

Author: Updated as suggested.

> <3> Optional: If you have configured the network interface with `nmstate`, and you want to disable an interface, set `state: up` with the IP addresses set to `enabled: false` as shown:

Author: Updated As Suggested.

+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-network-config-secret
  namespace: openshift-machine-api
 # ...
interfaces:
  - name: <nic_name>
    type: ethernet
    state: up
    ipv4:
      enabled: false
    ipv6:
      enabled: false
# ...
----
+

> <4> Replace `<nic1_name>`, `<ip_address>`, `<dns_ip_address>`, `<next_hop_ip_address>`, and `<next_hop_nic1_name>` with appropriate values.

Author: Updated as suggested.

<5> Replace `<base64_of_uid>` and  `<base64_of_pwd>` with the base64 string of the user name and password.
<6> Replace `<nic1_mac_address>` with the MAC address of the bare-metal node's first NIC. See the "BMC addressing" section for additional BMC configuration options.
<7> Replace `<protocol>` with the BMC protocol, such as IPMI, Redfish, or others. Replace `<bmc_url>` with the URL of the bare-metal node's baseboard management controller.

> <8> Optional: To skip TLS certificate validation, set `disableCertificateVerification` to `true`. Disabling TLS certificate validation is not recommended.

Author: Omitting may make the most sense here, because you're right that there's no real need to call it out if we don't want the user to do it. We can always update the content later if it becomes a question.

<9> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint.
--

* *DHCP configuration* `bmh.yaml`:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <2>
  password: <base64_of_pwd> <2>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: true
  bootMACAddress: <nic1_mac_address> <3>
  bmc:
    address: <protocol>://<bmc_url> <4>
    credentialsName: openshift-worker-<num>-bmc
    disableCertificateVerification: True <5>
  customDeploy:
    method: install_coreos
  userData:
    name: worker-user-data-managed
    namespace: openshift-machine-api
  rootDeviceHints:
    deviceName: <root_device_hint> <6>
----
+

> <1> Replace `<num>` with a unique worker number for the bare-metal nodes in the `name`, and `credentialsName` fields.

Author: Updated as suggested.

<2> Replace `<base64_of_uid>` and `<base64_of_pwd>` with the base64 string of the user name and password.
<3> Replace `<nic1_mac_address>` with the MAC address of the bare-metal node's first NIC. See the "BMC addressing" section for additional BMC configuration options.
<4> Replace `<protocol>` with the BMC protocol, such as IPMI, Redfish, or others. Replace `<bmc_url>` with the URL of the bare-metal node's baseboard management controller.

> <5> To skip certificate validation, set `disableCertificateVerification` to true.

Author: Removed optional step as [suggested](https://github.com/openshift/openshift-docs/pull/77264/files#r1786480221).

> <6> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint. See "Root device hints" for additional details.

Author: Updated as suggested.

+

> [NOTE]

Author: Updated As Suggested.

====
If the MAC address of an existing bare-metal node matches the MAC address of a bare-metal host that you are attempting to provision, then the installation will fail. If the host enrollment, inspection, cleaning, or other steps fail, the Bare Metal Operator retries the installation continuously. See "Diagnosing a host duplicate MAC address" for more information.
====

> . Create the bare-metal node:

Author: Updated as suggested.

. Create the `Secret` CR and the `BareMetalHost` CR.

. Create a configuration file for the bare-metal node. Depending on whether you are using a static configuration or a DHCP server, use one of the following example `bmh.yaml` files, replacing values in the YAML to match your environment:
+
[source,terminal]
----
$ vim bmh.yaml
----
* *Static configuration* `bmh.yaml`:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:

> name: openshift-worker-<num>-network-config-secret <1>
> namespace: openshift-machine-api

Reviewer 2: ```suggestion
  name: openshift-worker-<num>-network-config-secret <1>
  namespace: openshift-machine-api
```

Author: Updated As Suggested.

type: Opaque
stringData:
 nmstate: | <2>
  interfaces: <3>
  - name: <nic1_name> <4>
    type: ethernet
    state: up
    ipv4:
      address:
      - ip: <ip_address> <4>
        prefix-length: 24
      enabled: true
  dns-resolver:
    config:
      server:
      - <dns_ip_address> <4>
  routes:
    config:
    - destination: 0.0.0.0/0
      next-hop-address: <next_hop_ip_address> <4>
      next-hop-interface: <next_hop_nic1_name> <4>
---
apiVersion: v1
kind: Secret
metadata:

> name: openshift-worker-<num>-bmc-secret <1>
> namespace: openshift-machine-api

Reviewer 2: ```suggestion
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
```

Author: Updated As Suggested.

type: Opaque
data:

>  username: <base64_of_uid> <5>
>  password: <base64_of_pwd> <5>

Reviewer 2: ```suggestion
  username: <base64_of_uid> <5>
  password: <base64_of_pwd> <5>
```

Author: Updated As Suggested.

---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: <openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <6>
  bmc:
    address: <protocol>://<bmc_url> <7>
    credentialsName: openshift-worker-<num>-bmc-secret <1>
    disableCertificateVerification: True <8>
  customDeploy:
    method: install_coreos

>   online: true

Reviewer 2: Delete line, already added on line 79

Author: Updated As Suggested.

  userData:
    name: worker-user-data-managed
    namespace: openshift-machine-api
  rootDeviceHints:

>     deviceName: <root_device_hint> <9>

Reviewer 2: Missed adding in my previous comment. This should be indented same as rootDeviceHints
```
      preprovisioningNetworkDataName: openshift-worker-<num>-network-config-secret <1>
```

Author: Updated As Suggested.

----
+
--
<1> Replace `<num>` with a unique worker number for the bare-metal nodes in the `name`, `credentialsName`, and `preprovisioningNetworkDataName` fields.
<2> Add the NMState YAML syntax to configure the host interfaces. To configure the network interface for a newly created node, specify the name of the secret that has the network configuration. Follow the `nmstate` syntax to define the network configuration for your node. See "Optional: Configuring host network interfaces" for details on configuring NMState syntax.
<3> Optional: If you have configured the network interface with `nmstate`, and you want to disable an interface, set `state: up` with the IP addresses set to `enabled: false` as shown:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:

>  name: openshift-worker-<num>-network-config-secret
>  namespace: openshift-machine-api

Reviewer 2: ```suggestion
  name: openshift-worker-<num>-network-config-secret
  namespace: openshift-machine-api
```

<7> Replace `<protocol>` with the BMC protocol, such as IPMI, Redfish, or others. Replace `<bmc_url>` with the URL of the bare-metal node's baseboard management controller.
<8> Optional: To skip TLS certificate validation, set `disableCertificateVerification` to `true`. Disabling TLS certificate validation is not recommended.
<9> Optional: Replace `<root_device_hint>` with a device path if you specify a root device hint.
--

* *DHCP configuration* `bmh.yaml`:
+
[source,yaml]
----
---
apiVersion: v1
kind: Secret
metadata:
  name: openshift-worker-<num>-bmc-secret <1>
  namespace: openshift-machine-api
type: Opaque
data:
  username: <base64_of_uid> <2>
  password: <base64_of_pwd> <2>
---
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: openshift-worker-<num> <1>
  namespace: openshift-machine-api
spec:
  online: True
  bootMACAddress: <nic1_mac_address> <3>
  bmc:
    address: <protocol>://<bmc_url> <4>
    credentialsName: openshift-worker-<num>-bmc
    disableCertificateVerification: True <5>
  customDeploy:
    method: install_coreos

> online: true

Reviewer 2: ```suggestion
```
Online already added on line 149

Author: Updated As Suggested.

> userData:
>   name: worker-user-data-managed
>   namespace: openshift-machine-api
> rootDeviceHints:
>   deviceName: <root_device_hint> <6>

Reviewer 2: ```suggestion
  userData:
    name: worker-user-data-managed
    namespace: openshift-machine-api
  rootDeviceHints:
    deviceName: <root_device_hint> <6>
```

Author: Updated As Suggested.

.Example output
[source,terminal]
----
secret/openshift-worker-<num>-network-config-secret created
secret/openshift-worker-<num>-bmc-secret created
baremetalhost.metal3.io/openshift-worker-<num> created
----
+

> Where `<num>` will be the worker number.

Reviewer 8: Use present tense.

Author: Updated As Suggested.

.Example output
[source,terminal]
----
secret/openshift-worker-<num>-network-config-secret created
secret/openshift-worker-<num>-bmc-secret created
baremetalhost.metal3.io/openshift-worker-<num> created
----
+
Where `<num>` is the worker number.

> . Power on and inspect the bare-metal node:

Reviewer 2: The node being added should be in the power down state before running the step to create the bmh above. The node will be powered on when the create command is executed. It will move from states registering, then inspecting, provisioning and finally provisioned. if all goes well The progress of the states can be checked with running the below get bmh command in  regular intervals.

Author: Updated As Suggested.

+
[source,terminal]
----
$ oc -n openshift-machine-api get bmh openshift-worker-<num>
----
+
Where `<num>` is the worker node number.
+
.Example output
[source,terminal]
----
NAME                    STATE       CONSUMER   ONLINE   ERROR

> openshift-worker-<num>  available              true

Reviewer 2: Available state would only happen if either the customDeploy or userData section is missing in the yaml while creating the bmh. It also happens while [Removing hosts from a user-provisioned cluster by using the BMO](https://77264--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_bare_metal/scaling-a-user-provisioned-cluster-with-the-bare-metal-operator.html#upi-removing-hosts-from-a-upi-cluster_scaling-a-user-provisioned-cluster-with-the-bare-metal-operator) 
Step 3 should delete the bmh that is now in available state
`oc delete bmh -n openshift-machine-api <bmh_name>`

Step 4 will be delete node

In step 2
update the oc edit bmh command to run on  `<bmh_name>` as well

Author: Updated As Suggested.

----
+
[NOTE]
====

> To allow the worker node to join the cluster, scale the `machineset` object to the number of the `BareMetalHost` objects. You can scale nodes either manually or automatically. To scale nodes automatically, use the `metal3.io/autoscale-to-hosts` annotation for `machineset`.

Reviewer 2: This note is not applicable  in UPI cluster, as there are no Machineset to scale.

Author: Updated As Suggested.

.Example output
[source,terminal]
----
secret/openshift-worker-<num>-network-config-secret created
secret/openshift-worker-<num>-bmc-secret created
baremetalhost.metal3.io/openshift-worker-<num> created
----
+

> where:

Reviewer 4: up to you, but for a "where" that is a single item you can do it all in one line like:

```suggestion
where `<num>` specifies the worker node number.
```

Author: Updated as suggested.

<num>:: Specifies the worker node number.

. Inspect the bare-metal node by running the following command:
+
[source,terminal]
----
$ oc -n openshift-machine-api get bmh openshift-worker-<num>
----
+
where:

> <num>:: Specifies the worker node number.

Reviewer 4: ```suggestion
<num>:: Specifies the compute node number.
```

Author: Updated as suggested.

+

> .Example output

Reviewer 4: This is embedding under the description list instead of the step. If you want to use a desc list here, put it in an open block to fix the indent. (or you could single-line it as mentioned above, or maybe omit this "where" entirely since num is defined right before)

Author: Updated as suggested.

.Example output
[source,terminal]
----
secret/openshift-worker-<num>-network-config-secret created
secret/openshift-worker-<num>-bmc-secret created
baremetalhost.metal3.io/openshift-worker-<num> created
----
+
where:

<num>:: Specifies the worker node number.

. Inspect the bare-metal node by running the following command:
+
[source,terminal]
----
$ oc -n openshift-machine-api get bmh openshift-worker-<num>
----
+
where:

<num>:: Specifies the worker node number.
+
.Example output
[source,terminal]
----
NAME                    STATE       CONSUMER   ONLINE   ERROR

> openshift-worker-<num>  provisioned true

Reviewer 5: I cannot comment on line 225, but there you have the old BMH name. Probably you don't need to check that the state is "provisioned" for the 2nd time.

Author: Updated as suggested.

.Example output
[source,terminal]
----
secret/openshift-worker-<num>-network-config-secret created
secret/openshift-worker-<num>-bmc-secret created
baremetalhost.metal3.io/openshift-worker-<num> created
----
+

> Where `<num>` is the worker number.

Author: Updated as suggested.

> . Inspect the bare-metal node:

Author: Updated as suggested.

+
[source,terminal]
----
$ oc -n openshift-machine-api get bmh openshift-worker-<num>
----
+

> Where `<num>` is the worker node number.

Author: Updated as suggested.
